gazebo/sensors/SensorManager.cc

```diff
@@ -132,6 +172,130 @@ void SensorManager::WaitForSensors(double _clk, double _dt)
   }
 }
 
+//////////////////////////////////////////////////
+void PublishPerformanceMetrics()
+{
+  if (node == nullptr)
+  {
+    auto world = physics::get_world();
+    // Transport
+    node = transport::NodePtr(new transport::Node());
+    node->Init(world->Name());
+    performanceMetricsPub =
+     node->Advertise<msgs::PerformanceMetrics>(
+         "/gazebo/performance_metrics", 10, 5);
+  }
+
+  if (!performanceMetricsPub || !performanceMetricsPub->HasConnections())
+  {
+    return;
+  }
+
+  physics::WorldPtr world = physics::get_world(
+    gazebo::physics::get_world()->Name());
+
+  /// Outgoing run-time simulation performance metrics.
+  msgs::PerformanceMetrics performanceMetricsMsg;
+
+  // Real time factor
+  common::Time realTime = world->RealTime();
+  common::Time diffRealtime = realTime - lastRealTime;
+  common::Time simTime = world->SimTime();
+  common::Time diffSimTime = simTime - lastSimTime;
+  common::Time realTimeFactor;
+
+  if (ignition::math::equal(diffSimTime.Double(), 0.0))
+    return;
+
+  if (realTime == 0)
+    realTimeFactor = 0;
+  else
+    realTimeFactor = diffSimTime / diffRealtime;
+
+  if (realTimeFactor > 0)
+    performanceMetricsMsg.set_real_time_factor(realTimeFactor.Double());
+  else
+    performanceMetricsMsg.set_real_time_factor(0.0);
+
+  lastRealTime = realTime;
+  lastSimTime = simTime;
+
+  /// update sim time for sensors
+  for (auto model: world->Models())
+  {
+    for (auto link: model->GetLinks())
+    {
+      for (unsigned int i = 0; i < link->GetSensorCount(); i++)
+      {
+        std::string name = link->GetSensorName(i);
+        sensors::SensorPtr sensor = sensors::get_sensor(name);
+
+        auto ret = sensorsLastMeasurementTime.insert(
+            std::pair<std::string, gazebo::common::Time>(name, 0));
+        worldLastMeasurementTime.insert(
+                std::pair<std::string, gazebo::common::Time>(name, 0));
+        if (ret.second == false)
+        {
+          double updateSimRate =
+            (sensor->LastMeasurementTime() - sensorsLastMeasurementTime[name])
+            .Double();
+          if (updateSimRate > 0.0)
+          {
+            sensorsLastMeasurementTime[name] = sensor->LastMeasurementTime();
+            double updateRealRate =
+              (world->RealTime() - worldLastMeasurementTime[name]).Double();
+
+            struct sensorPerformanceMetricsType emptySensorPerfomanceMetrics;
+            auto ret2 = sensorPerformanceMetrics.insert(
+                std::pair<std::string, struct sensorPerformanceMetricsType>
+                  (name, emptySensorPerfomanceMetrics));
+            if (ret2.second == false)
+            {
+              sensorPerformanceMetrics[name].sensorSimUpdateRate =
+                  1.0/updateSimRate;
+              sensorPerformanceMetrics[name].sensorRealUpdateRate =
+                  1.0/updateRealRate;
```

Comment:
You can avoid indexing into the map by reusing iterator returned by `insert()`.   Not sure if it'll matter performance wise but it doesn't hurt readability doing so regardless IMO.  Same below for the avg FPS.